File: springy-java/src/net/jangxyz/springy/Vector.java
package net.jangxyz.springy;

public class Vector {
	public double x;
	public double y;

    public static Vector ZERO = new Vector(0, 0);
	public static Vector random() {
		return new Vector(10.0 * (Math.random() - 0.5), 10.0 * (Math.random() - 0.5));
	}


	public Vector(double x, double y) {
		this.x = x;
		this.y = y;
	}
	
	@Override
	public String toString() {
		return String.format("(%f, %f)", this.x, this.y);		
	}
	public String toString(int precision) {
		String formatStr = String.format("(%%.%df, %%.%df)", precision, precision);
		return String.format(formatStr, this.x, this.y);		
	}

	//
	// operations
	//
	public Vector add(Vector v) {
		return new Vector(this.x + v.x, this.y + v.y);
	}

	public Vector subtract(Vector v) {
		return new Vector(this.x - v.x, this.y - v.y);
	}

	public Vector multiply(double n) {
		return new Vector(this.x * n, this.y * n);
	}

	public Vector divide(double n) {
		// Avoid divide by zero errors..
		return new Vector((this.x / n), (this.y / n));
	}
	
	public double magnitude2() {
		return this.x*this.x + this.y*this.y;
	}

	public double magnitude() {
		return Math.sqrt(this.x*this.x + this.y*this.y);
	}

	public Vector normal() {
		return new Vector(-this.y, this.x);
	}

	public Vector normalise() {
		return this.divide(this.magnitude());
	}
	
	public boolean equals(Vector v) {
		return this.x == v.x && this.y == v.y;
	}

}

File: springy-swift/springy/Vector.swift

//
//  vector.swift
//  springy-swift
//
//  Created by 김장환 on 1/11/15.
//  Copyright (c) 2015 김장환. All rights reserved.
//

import Foundation

let ARC4_RANDOM_MAX = Double(0x100000000)
func rand() -> Double
{
    var randDouble = Double(arc4random())
    return randDouble / ARC4_RANDOM_MAX
}

struct Vector: Equatable, Printable {
    let x: Double
    let y: Double
    let magnitude2: Double
    let magnitude:  Double
    
    static func random() -> Vector {
        return Vector(x:rand(), y:rand())
    }
    
    static var ZERO: Vector {
        return Vector(x:0, y:0)
    }
    
    init(x: Double, y: Double) {
        self.x = x
        self.y = y
        
        self.magnitude2 = self.x * self.x + self.y * self.y
        self.magnitude = sqrt(self.magnitude2)
    }

    var description: String {
        return "Vector:(\(x),\(y))"
    }
    
    func add(v2: Vector) -> Vector {
        return Vector(x:self.x + v2.x, y:self.y + v2.y)
    }
    
    func subtract(v2: Vector) -> Vector {
        return Vector(x:self.x - v2.x, y:self.y - v2.y)
    }
    
    func multiply(n: Double) -> Vector {
        return Vector(x:self.x * n, y:self.y * n)
    }
    
    func divide(n: Double) -> Vector {
        return Vector(x:(self.x / n), y:(self.y / n)) // Avoid divide by zero errors..
    }
    
    func normal() -> Vector {
        return Vector(x: -self.y, y: self.x)
    }

    func normalise() -> Vector {
        return self / self.magnitude
    }
    func normalize() -> Vector {
        return self.normalise()
    }
    
}

func == (left: Vector, right: Vector) -> Bool {
    return (left.x == right.x) && (left.y == right.y)
}
func != (left: Vector, right: Vector) -> Bool {
    return !(left == right)
}

func + (left: Vector, right: Vector) -> Vector {
    return left.add(right)
}
func - (left: Vector, right: Vector) -> Vector {
    return left.subtract(right)
}
func * (v: Vector, n: Double) -> Vector {
    return v.multiply(n)
}
func * (n:Double, v: Vector) -> Vector {
    return v.multiply(n)
}
func / (v: Vector, n: Double) -> Vector {
    return v.divide(n)
}

func += (inout left: Vector, right: Vector) {
    left = left + right
}
